### PR TITLE
bug(rhineng-9037): Perform tag API functionality in python

### PR DESCRIPTION
**Pull request by chambridge**
# Overview

This PR is being created to address [RHINENG-9037](https://issues.redhat.com/browse/RHINENG-9037).
* Reduce burden on DB as current implementation is impacting DB performance
* Perform the search and pagination functionality in python code

## PR Checklist

- [x] Keep PR title short, ideally under 72 characters
- [ ] Descriptive comments provided in complex code blocks
- [ ] Tests: validate optimal/expected output
- [ ] Tests: validate exceptions and failure scenarios
- [ ] Tests: edge cases
- [ ] Recovers or fails gracefully during potential resource outages (e.g. DB, Kafka)
- [ ] Uses [type hinting](https://docs.python.org/3/library/typing.html), if convenient
- [ ] Documentation, if this PR changes the way other services interact with host inventory
- [ ] Links to related PRs

## Secure Coding Practices Documentation Reference

You can find documentation on this checklist [here](https://github.com/RedHatInsights/secure-coding-checklist).

## Secure Coding Checklist

- [ ] Input Validation
- [ ] Output Encoding
- [ ] Authentication and Password Management
- [ ] Session Management
- [ ] Access Control
- [ ] Cryptographic Practices
- [ ] Error Handling and Logging
- [ ] Data Protection
- [ ] Communication Security
- [ ] System Configuration
- [ ] Database Security
- [ ] File Management
- [ ] Memory Management
- [ ] General Coding Practices

### CODE REVIEW

**Review by kruai — CHANGES_REQUESTED** [review]
Thanks for hopping on this so fast. Got a couple of suggestions - also needs some DRY, but I think I introduced the duplication, so we can refactor after we resolve the DB resource usage issue

File: api/host_query_db.py [hunk 1]
```diff
@@ -188,11 +190,12 @@ def get_host_tags_list_by_id_list(
     all_filters += rbac_permissions_filter(rbac_filter)
     order = params_to_order_by(order_by, order_how)
     query_results = query.filter(*all_filters).order_by(*order).offset(offset).limit(limit).all()
-    host_tags_dict = _expand_host_tags(query_results)
+    host_tags_dict, _ = _expand_host_tags(query_results)
     return host_tags_dict, len(host_id_list)
 
 
 def _expand_host_tags(hosts: List[Host]) -> dict:
```
[the remark this answers is not in the document]
Comment (chambridge) [reply api/host_query_db.py]: fixed.

File: api/host_query_db.py [hunk 2]
```diff
@@ -201,39 +204,31 @@ def _expand_host_tags(hosts: List[Host]) -> dict:
...
                 if isinstance(tag_values, List):
                     for tag_value in tag_values:
+                        host_tag_str = f"{_convert_null_string(host_namespace)}/{_convert_null_string(tag_key)}={_convert_null_string(tag_value)}"  # noqa: E501
                         host_tag_obj = {
                             "namespace": _convert_null_string(host_namespace),
                             "key": _convert_null_string(tag_key),
                             "value": _convert_null_string(tag_value),
                         }
                         host_tags.append(host_tag_obj)
+                        if not host_tags_tracker.get(host_tag_str):
```
[the remark this answers is not in the document]
Comment (chambridge) [reply api/host_query_db.py]: fixed.

File: api/host_query_db.py [hunk 3]
```diff
@@ -201,39 +204,31 @@ def _expand_host_tags(hosts: List[Host]) -> dict:
...
-                        }
-                    )
+                    host_tag_str = f"{_convert_null_string(host_namespace)}/{_convert_null_string(tag_key)}={_convert_null_string(tag_values)}"  # noqa: E501
+                    host_tag_obj = {
+                        "namespace": _convert_null_string(host_namespace),
+                        "key": _convert_null_string(tag_key),
+                        "value": _convert_null_string(tag_values),
+                    }
+                    host_tags.append(host_tag_obj)
+                    if not host_tags_tracker.get(host_tag_str):
```
[the remark this answers is not in the document]
Comment (chambridge) [reply api/host_query_db.py]: fixed.

**Review by kruai — APPROVED** [review]

**Pull request merged**